### ISSUE TITLE
Any Examples with Autofac and ASP.NET MVC?

### ISSUE THREAD

**Issue by accoleon** [eventflow/EventFlow#318]
I'm unable to get a simple example working in the context of Autofac and ASP.NET MVC:

Inside Global.asax.cs.Application_Start():
```csharp
var containerBuilder = new ContainerBuilder();
containerBuilder.RegisterType<HomeController>().InstancePerLifetimeScope();

var webContainer = EventFlowOptions.New
    .UseAutofacContainerBuilder(containerBuilder)
    .AddEvents(typeof(ExampleEvent))
    .AddCommands(typeof(ExampleCommand))
    .AddCommandHandlers(typeof(ExampleCommandHandler))
    .UseInMemoryReadStoreFor<ExampleReadModel>()
    .CreateContainer();

DependencyResolver.SetResolver(new AutofacDependencyResolver(webContainer));
```
Inside HomeController.cs:
```csharp
public class HomeController : Controller
{
    private readonly ICommandBus _commandBus;
    private readonly IQueryProcessor _queryProcessor;
    public HomeController(ICommandBus commandBus, IQueryProcessor queryProcessor)
    {
        _commandBus = commandBus;
        _queryProcessor = queryProcessor;
    }

    public async Task<ViewResult> Index()
    {
        ViewBag.Title = "Home Page";
            const int magicNumber = 42;
            var exampleId = ExampleId.NewComb();
            await _commandBus.PublishAsync(new ExampleCommand(exampleId, magicNumber), CancellationToken.None);

            var readModel = await _queryProcessor.ProcessAsync(new ReadModelByIdQuery<ExampleReadModel>(exampleId), CancellationToken.None);
            return View(readModel);
    }
}
```
Inside Index.cshtml:
```html
@model ExampleReadModel
<h3>Server Read Model</h3>
<p>@Model.ExampleId</p>
<p>@Model.MagicNumber</p>
```

Usually the first load works fine, but once you try to reload the page, there will be a "The semaphore has been disposed." exception:
![image](https://cloud.githubusercontent.com/assets/115630/25421578/63d45930-2a12-11e7-9051-46c635d23dc1.png)

I've tried registering Autofac using specific scopes, but it still gives a similar exception (solution from http://stackoverflow.com/a/16840451)

Any idea or examples/gist on a simple end-to-end example in ASP.NET? None of the examples actually use eventflow in an ASP.NET MVC/WebAPI application and the associated boilerplate that Autofac requires to integrate with MVC/WebAPI.

**Comment by rasmus** [comment]
Hi @accoleon. Thanks for reporting this. I'll have to look into it, but it looks like a problem in the `InMemoryReadStore` which I'll have to look into when I have access to my computer. Could you try to remove it and verify that it indeed is the cause?

**Comment by kthe** [comment]
I have seen that error before.
That was inside an httppost. The solution was to use actionresult as a
retun type instead. Like..

public async Task<ActionResult> Index(..

/Krister


onsdag 26 april 2017 skrev Kevin Xu <notifications@github.com>:

**Comment by kthe** [comment]
After double checking, it turned out I'm using

[HttpPost]
        [Route("api/v1/schedule")]
        public async Task<IHttpActionResult> Process(
            ProcessProductionScheduleType schedule)
        {
            // perform the actual processing and
            // publish internal events
            await _productionScheduleService.ProcessSchedule(
                schedule,
                CancellationToken.None)
                .ConfigureAwait(false);

            return Ok();
        }


Not using IHttpActionResult gave me the same error.


onsdag 26 april 2017 skrev Zoomware <zoomware@gmail.com>:

**Comment by accoleon** [comment]
So even after removing the InMemoryReadStore:
```csharp
var containerBuilder = new ContainerBuilder();
containerBuilder.RegisterType<HomeController>().InstancePerLifetimeScope();

var webContainer = EventFlowOptions.New
    .UseAutofacContainerBuilder(containerBuilder)
    .AddEvents(typeof(ExampleEvent))
    .AddCommands(typeof(ExampleCommand))
    .AddCommandHandlers(typeof(ExampleCommandHandler))
    //.UseInMemoryReadStoreFor<ExampleReadModel>()
    .CreateContainer();

DependencyResolver.SetResolver(new AutofacDependencyResolver(webContainer));
```
 and my HomeController.cs looks like this:
```csharp
public class HomeController : Controller
{
    private readonly ICommandBus _commandBus;
    private readonly IQueryProcessor _queryProcessor;
    private readonly IAggregateStore _aggregateStore;

    public HomeController(ICommandBus commandBus, IQueryProcessor queryProcessor, IAggregateStore aggregateStore)
    {
        _commandBus = commandBus;
        _queryProcessor = queryProcessor;
        _aggregateStore = aggregateStore;
    }

    public async Task<ViewResult> Index()
    {
        ViewBag.Title = "Home Page";
        const int magicNumber = 42;
        var exampleId = ExampleId.New;
        await _commandBus.PublishAsync(new ExampleCommand(exampleId, magicNumber), CancellationToken.None);
        var aggregate = await _aggregateStore.LoadAsync<ExampleAggregate,ExampleId>(exampleId,CancellationToken.None);
        return View(aggregate);
    }
}
```
Index.cshtml:
```html
@model ExampleAggregate
<h3>Server Read Model</h3>
<p>@Model.Id</p>
<p>@Model.MagicNumber</p>
```
It's still the semaphore disposed error. I am pretty sure it has something to do with EventFlow's Scope Lifetime being custom and not what Autofac's MVC integration expects (and probably nothing to do with the InMemoryReadStore), but since it was stated that this has been in use in production (and hopefully in some ASP.NET MVC sites) that I can find a working example of a boilerplate setup of EventFlow/Autofac/MVC.

**Comment by rasmus** [comment]
@accoleon I created a MVC5 project in #319 with the code you stated in the very beginning. However, it doesn't fail. Could you have a look and see what's different?

**Comment by accoleon** [comment]
I checked out that branch and it runs fine the first time, but reloading the same page causes the semaphore disposed error - you don't experience it?

**Comment by rasmus** [comment]
@accoleon I do... I found the bug. Its a problem with the singleton registrations. I'm looking into how the test `TestSuiteForServiceRegistration.Singletons` didn't catch it. I'll release a fix as soon as its done

**Comment by rasmus** [comment]
@accoleon and @kthe found the bug. `IDisposable.Dispose()` was invoked on singletons due to a bug in EventFlow service registration that enables easier decoration.

**Comment by rasmus** [comment]
Fixed in #320

**Comment by rasmus** [comment]
@accoleon Could you verify the fix in 0.41.2727?

**Comment by accoleon** [comment]
Yes, I updated the EventFlow and EventFlow.Autofac NuGet packages to 0.41.2727 and now the test project builds fine works across refreshes. Thanks for fixing the bug quickly!
